components/ws-proxy/pkg/ratelimit/rate-limited-listener.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package ratelimit
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"golang.org/x/time/rate"
+)
+
+// NewRateLimitedListen returns a net.Listen implementation that adds rate limiting.
+func NewRateLimitedListen(ctx context.Context, refillInterval time.Duration, bucketSize int) func(network, address string) (net.Listener, error) {
+	return func(network, address string) (net.Listener, error) {
+		return newRateLimitedListener(
+			ctx,
+			network,
+			address,
+			refillInterval,
+			bucketSize,
+		)
```

Comment:
nit: why the indirection? Couldn't the code in `newRateLimitedListener` be here directly?

Reply:
Right. It's a left-over from code restructuring.